[JIT] fix common_expression_hoisting

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74924
* __->__ #74794

aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

Differential Revision: [D35188927](https://our.internmc.facebook.com/intern/diff/D35188927)

davidberard98 committed:
[JIT] fix common_expression_hoisting

aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] fix common_expression_hoisting"


aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] fix common_expression_hoisting"


aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).

davidberard98 commented:
Turns out previous fix is insufficient, here's a new repro that still fails

```python
import torch
from typing import Dict

def fn(x, y, cond: bool, d: Dict[str, torch.Tensor]):
    if cond:
        m = x.relu()
        f1 = torch.rand((2, 2))
        d["test"] = f1
        z = d["test"]
    else:
        m = y.gelu()
        f2 = torch.rand((3, 2))
        d["test"] = f2
        z = d["test"]
    return m, z

fn_s = torch.jit.script(fn)
x = torch.rand((2, 2))
y = torch.rand((2, 2))
d = {"x": x}
fn_s(x, y, 1, d)
```

Gamrix commented:
The fix looks good, but there seems to be a second bug that is unrelated to this change that you are catching with your new repro unless GraphExecutor is running optimizations on the first graph run. 

Print of the graph right before the `fn_s(...)` call in your repro:
```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %cond.1 : bool,
      %d.1 : Dict(str, Tensor)):
  %24 : str = prim::Constant[value="none"]()
  %18 : str = prim::Constant[value="test"]() # <ipython-input-2-bf373c45d473>:8:10
  %11 : NoneType = prim::Constant()
  %8 : int = prim::Constant[value=2]() # <ipython-input-2-bf373c45d473>:7:25
  %26 : int = prim::Constant[value=3]() # <ipython-input-2-bf373c45d473>:12:25
  %m : Tensor, %z : Tensor = prim::If(%cond.1) # <ipython-input-2-bf373c45d473>:5:4
    block0():
      %m.1 : Tensor = aten::relu(%x.1) # <ipython-input-2-bf373c45d473>:6:12
      %10 : int[] = prim::ListConstruct(%8, %8)
      %f1.1 : Tensor = aten::rand(%10, %11, %11, %11, %11) # <ipython-input-2-bf373c45d473>:7:13
       = aten::_set_item(%d.1, %18, %f1.1) # <ipython-input-2-bf373c45d473>:8:8
      %z.1 : Tensor = aten::__getitem__(%d.1, %18) # <ipython-input-2-bf373c45d473>:9:12
      -> (%m.1, %z.1)
    block1():
      %m.3 : Tensor = aten::gelu(%y.1, %24) # <ipython-input-2-bf373c45d473>:11:12
      %28 : int[] = prim::ListConstruct(%26, %8)
      %f2.1 : Tensor = aten::rand(%28, %11, %11, %11, %11) # <ipython-input-2-bf373c45d473>:12:13
       = aten::_set_item(%d.1, %18, %f2.1) # <ipython-input-2-bf373c45d473>:13:8
      %z.3 : Tensor = aten::__getitem__(%d.1, %18) # <ipython-input-2-bf373c45d473>:14:12
      -> (%m.3, %z.3)
  %51 : (Tensor, Tensor) = prim::TupleConstruct(%m, %z)
  return (%51)
  ```

Gamrix commented:
From your repo, `aten::_set_item` is somehow not being treated as aliasing  `aten::_getitem__` 

```
[DUMP common_expression_hoisting.cpp:148] After CEH Changes
[DUMP common_expression_hoisting.cpp:148] graph(%x.1 : Tensor,
[DUMP common_expression_hoisting.cpp:148]       %y.1 : Tensor,
[DUMP common_expression_hoisting.cpp:148]       %cond.1 : bool,
[DUMP common_expression_hoisting.cpp:148]       %d.1 : Dict(str, Tensor)):
[DUMP common_expression_hoisting.cpp:148]   %6 : NoneType = prim::Constant()
[DUMP common_expression_hoisting.cpp:148]   %7 : str = prim::Constant[value="test"]() # misc/ceh_bug_repro.py:9:10
[DUMP common_expression_hoisting.cpp:148]   %8 : str = prim::Constant[value="none"]()
[DUMP common_expression_hoisting.cpp:148]   %z.3 : Tensor = aten::__getitem__(%d.1, %7) # misc/ceh_bug_repro.py:15:12
[DUMP common_expression_hoisting.cpp:148]   %m : Tensor = prim::If(%cond.1) # misc/ceh_bug_repro.py:6:4
[DUMP common_expression_hoisting.cpp:148]     block0():
[DUMP common_expression_hoisting.cpp:148]       %m.1 : Tensor = aten::relu(%x.1) # misc/ceh_bug_repro.py:7:12
[DUMP common_expression_hoisting.cpp:148]       %20 : int[] = prim::Constant[value=[2, 2]]()
[DUMP common_expression_hoisting.cpp:148]       %f1.1 : Tensor = aten::rand(%20, %6, %6, %6, %6) # misc/ceh_bug_repro.py:8:13
[DUMP common_expression_hoisting.cpp:148]        = aten::_set_item(%d.1, %7, %f1.1) # misc/ceh_bug_repro.py:9:8
[DUMP common_expression_hoisting.cpp:148]       -> (%m.1)
[DUMP common_expression_hoisting.cpp:148]     block1():
[DUMP common_expression_hoisting.cpp:148]       %m.3 : Tensor = aten::gelu(%y.1, %8) # misc/ceh_bug_repro.py:12:12
[DUMP common_expression_hoisting.cpp:148]       %21 : int[] = prim::Constant[value=[3, 2]]()
[DUMP common_expression_hoisting.cpp:148]       %f2.1 : Tensor = aten::rand(%21, %6, %6, %6, %6) # misc/ceh_bug_repro.py:13:13
[DUMP common_expression_hoisting.cpp:148]        = aten::_set_item(%d.1, %7, %f2.1) # misc/ceh_bug_repro.py:14:8
[DUMP common_expression_hoisting.cpp:148]       -> (%m.3)
[DUMP common_expression_hoisting.cpp:148]   %19 : (Tensor, Tensor) = prim::TupleConstruct(%m, %z.3)
[DUMP common_expression_hoisting.cpp:148]   return (%19)
```

davidberard98 committed:
Update on "[JIT] fix common_expression_hoisting"


aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

Differential Revision: [D35188927](https://our.internmc.facebook.com/intern/diff/D35188927)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] fix common_expression_hoisting"


aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

Differential Revision: [D35188927](https://our.internmc.facebook.com/intern/diff/D35188927)

[ghstack-poisoned]

davidberard98 commented:
@Gamrix `couldMoveBeforeTopologicallyValid` returns true if it's possible to move the specified node and some "[working set](https://github.com/pytorch/pytorch/blob/8ed6cb42ba2635319b67417b9ee2f1a0142d2c25/torch/csrc/jit/ir/alias_analysis.cpp#L1708)" to before the target node

In the case of the 2nd repro, we try to move the `getitem` node. It's possible to move this all before the first node as long as we include the `set_item` node in the working set, so `couldMoveBeforeTopologicallyValid` returns true.

So then we were moving `getitem` outside of the if block, even if we require `setattr` to appear before it.

davidberard98 committed:
Update on "[JIT] fix common_expression_hoisting"


aliasdb check was incorrect, caused issues in #74056

This checks whether it is safe to move before the node at the beginning
of the if branch instead of after the node at the beginning of the if
branch

Differential Revision: [D35188927](https://our.internmc.facebook.com/intern/diff/D35188927)

[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/passes/common_expression_hoisting.cpp:
```diff
@@ -42,18 +42,34 @@ struct CommonExpressionHoister {
+      // node. If it ended up at the very beginning of the if block, then it's
+      // safe to move it outside.
+      if (true_b_node != true_block->nodes().front() ||
+          false_b_node != false_block->nodes().front()) {
```
I think it's guaranteed that n_{i+1}, ... , n_i are unaffected by this comment https://github.com/pytorch/pytorch/blob/9872a06d77582e91e834103db75f774ca75f7fff/torch/csrc/jit/ir/alias_analysis.h#L115-L123

Gamrix replied:
You're right, I forgot that we already incremented the iterator before we did all the moves.

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35188927).
